### PR TITLE
Enable garbage collector for replicaset and rc controller

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://github.com/openshift/origin/issues/12927

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
@kargakis @smarterclayton @deads2k PTAL

(guess this should be backported to release-1.5)

File: pkg/cmd/server/kubernetes/master.go [hunk 1]
```diff
@@ -192,7 +192,11 @@ func (c *MasterConfig) RunReplicaSetController(client *kclientset.Clientset) {
 		client,
 		replicasetcontroller.BurstReplicas,
 		int(c.ControllerManager.LookupCacheSizeForRC),
-		c.ControllerManager.EnableGarbageCollector,
+		// FIXME: This should be set to c.ControllerManager.EnableGarbageCollector, however
```
Comment (kargakis) [inline pkg/cmd/server/kubernetes/master.go]: This knob is going away in 1.6, not sure the comment is necessary.

**Comment by mfojtik** [comment]
For the reference https://github.com/openshift/origin/blob/10760db7b556afe8949334501fe6fbb33cd4b828/vendor/k8s.io/kubernetes/pkg/controller/replicaset/replica_set.go#L601

🍺 to @kargakis for this catch

File: pkg/cmd/server/kubernetes/master.go (hunk 1 above)
Comment (mfojtik) [inline pkg/cmd/server/kubernetes/master.go]: rephrased.

**Comment by smarterclayton** [comment]
lgtm [merge]

**Comment by kargakis** [comment]
LGTM

We may want to verify that owner references are indeed set before merging - maybe a check in e2e.

**Comment by mfojtik** [comment]
@kargakis can you check the test failures? smells weird.

**Comment by kargakis** [comment]
Can't check right now but I think you also need to update the controller roles to patch pods.

**Comment by kargakis** [comment]
And add a check in e2e please

**Comment by smarterclayton** [comment]
Or turn on the garbage collection e2e test for these resources.

**Comment by kargakis** [comment]
The gc controller is not running

**Comment by smarterclayton** [comment]
We should probably turn it on in master unless we have a reason.

**Comment by mfojtik** [comment]
[test]

File: pkg/cmd/server/bootstrappolicy/infra_sa_policy.go [hunk 2]
```diff
@@ -243,7 +243,7 @@ func init() {
 					APIGroups: []string{""},
 					// TODO: remove "update" once
 					// https://github.com/kubernetes/kubernetes/issues/36897 is resolved.
-					Verbs:     sets.NewString("get", "list", "watch", "update"),
+					Verbs:     sets.NewString("get", "list", "watch", "update", "patch"),
```
Comment (kargakis) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: Why? Deployments don't adopt pods but replicasets. Added in 1.6.

Comment (mfojtik) [inline pkg/cmd/server/bootstrappolicy/infra_sa_policy.go]: thought deployments also patching pods... will nuke this

**Comment by kargakis** [comment]
> We should probably turn it on in master unless we have a reason.

@deads2k would know better

**Comment by deads2k** [comment]
> We should probably turn it on in master unless we have a reason.
> @deads2k would know better

Are we talking about for openshift 1.5 or 1.6?  The implementation is likely to be different between the two.  I'd try the 1.6 one (which we really need), I don't know that I'd do the 1.5 one.  It's failure mode is really bad.

**Comment by kargakis** [comment]
@smarterclayton was talking about master but master right now has kube 1.5 so we may want a post-rebase issue.

**Comment by mfojtik** [comment]
definitely a post-rebase issue, this is trying to address 1.5 bug before we close 1.5 release... 

@kargakis it seems like adding PATCH to allowed actions does not help and the kubectl rolling update is still stuck :-/ any idea what could be wrong there? 

I added extended test that exercise both RS and RC to make sure they only manage their pods (when both use same selectors).

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
@kargakis I reproduce the failure locally... after `kubectl rolling-update` I can see 2 pods instead of 1. The first pod has ownerReference set to "old" replication controller (UUID) and the second pod is owned by the "new" replication controller...

That seems like a bug in `kubectl rollling-update` or I'm missing something?

**Comment by mfojtik** [comment]
@kargakis i'm thinking about disabling those tests until we have GC enabled. If we have to backport this back to 1.4, I'm really hasitate to enable GC. In that case I would rather prefer to leave this as it is until 1.6.

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
We decided to opt-out on this fix for 3.5 and instead enable the 1.6 garbage collector properly for 3.6

**Pull request closed without merging**